go/pkg/bertymessenger/handlers.go

```diff
@@ -1,59 +1,129 @@
+		bertytypes.EventTypeGroupMetadataPayloadSent:              h.groupMetadataPayloadSent,
+		bertytypes.EventTypeAccountServiceTokenAdded:              h.accountServiceTokenAdded,
+	}
+
+	h.appMessageHandlers = map[AppMessage_Type]struct {
+		handler        func(tx *dbWrapper, i *Interaction, amPayload proto.Message) (*Interaction, error)
+		isVisibleEvent bool
+	}{
+		AppMessage_TypeAcknowledge:     {h.handleAppMessageAcknowledge, false},
+		AppMessage_TypeGroupInvitation: {h.handleAppMessageGroupInvitation, true},
+		AppMessage_TypeUserMessage:     {h.handleAppMessageUserMessage, true},
+		AppMessage_TypeSetUserName:     {h.handleAppMessageSetUserName, false},
+	}
+
+	return h
+}
+
+func (h *eventHandler) handleMetadataEvent(gme *bertytypes.GroupMetadataEvent) error {
 	et := gme.GetMetadata().GetEventType()
-	svc.logger.Info("received protocol event", zap.String("type", et.String()))
-
-	var err error
-	switch et {
-	case bertytypes.EventTypeAccountGroupJoined:
-		err = accountGroupJoined(svc, gme)
-	case bertytypes.EventTypeAccountContactRequestOutgoingEnqueued:
-		err = accountContactRequestOutgoingEnqueued(svc, gme)
-	case bertytypes.EventTypeAccountContactRequestOutgoingSent:
-		err = accountContactRequestOutgoingSent(svc, gme)
-	case bertytypes.EventTypeAccountContactRequestIncomingReceived:
-		err = accountContactRequestIncomingReceived(svc, gme)
-	case bertytypes.EventTypeAccountContactRequestIncomingAccepted:
-		err = accountContactRequestIncomingAccepted(svc, gme)
-	case bertytypes.EventTypeGroupMemberDeviceAdded:
-		err = groupMemberDeviceAdded(svc, gme)
-	case bertytypes.EventTypeGroupMetadataPayloadSent:
-		err = groupMetadataPayloadSent(svc, gme)
-	case bertytypes.EventTypeAccountServiceTokenAdded:
-		err = accountServiceTokenAdded(svc, gme)
-	default:
-		svc.logger.Info("event ignored", zap.String("type", et.String()))
+	h.logger.Info("received protocol event", zap.String("type", et.String()))
+
+	handler, ok := h.metadataHandlers[et]
+
+	if !ok {
+		h.logger.Info("event ignored", zap.String("type", et.String()))
+		return nil
 	}
 
-	return err
+	return handler(gme)
+}
+
+func (h *eventHandler) handleAppMessage(gpk string, gme *bertytypes.GroupMessageEvent, am *AppMessage) error {
+	h.logger.Info("handling app message", zap.String("type", am.GetType().String()))
+
+	// build interaction
+	i, err := interactionFromAppMessage(h, gpk, gme, am)
+	if err != nil {
+		return err
+	}
+
+	handler, ok := h.appMessageHandlers[i.GetType()]
+
+	if !ok {
+		h.logger.Warn("unsupported app message type", zap.String("type", i.GetType().String()))
+
+		return nil
+	}
+
+	// start a transaction
+	return h.db.tx(func(tx *dbWrapper) error {
+		if err := h.hydrateInteraction(gme, i); err != nil {
+			return err
+		}
+
+		// parse payload
+		amPayload, err := am.UnmarshalPayload()
+		if err != nil {
+			return err
+		}
+
+		i, err := handler.handler(tx, i, amPayload)
```

Comment:
```suggestion
		i, err = handler.handler(tx, i, amPayload)
```

it may be expected, if so, please add a comment explaining why